internal/providers/openstack/openstack.go

```diff
@@ -166,14 +173,78 @@ func fetchConfigFromDevice(logger *log.Logger, ctx context.Context, path string)
 	return os.ReadFile(filepath.Join(mnt, configDriveUserdataPath))
 }
 
+// checks if the given IP is an IPv6 address.
+func isIPv6Address(ip net.IP) bool {
+	isIPv6 := ip.To4() == nil
+	return isIPv6
+}
+
+// findNetworkInterfaceWithIPv6 returns the name of the first network interface with an active IPv6 address.
+// This interface name is needed to format the link-local address for accessing the IPv6 metadata service.
+// For more details, see: https://docs.openstack.org/nova/2024.1/user/metadata.html
+func findNetworkInterfaceWithIPv6() error {
+	interfaces, err := net.Interfaces()
+	if err != nil {
+		return fmt.Errorf("error fetching network interfaces: %v", err)
+	}
+
+	for _, iface := range interfaces {
+
+		if iface.Flags&net.FlagUp == 0 || iface.Flags&net.FlagLoopback != 0 {
+			continue
+		}
+
+		addrs, err := iface.Addrs()
+		if err != nil {
+			fmt.Printf("Error fetching addresses for interface %s: %v\n", iface.Name, err)
+			continue
+		}
+
+		for _, addr := range addrs {
+			if ipnet, ok := addr.(*net.IPNet); ok && isIPv6Address(ipnet.IP) {
+				// Prepare the IPv6 metadata service URL
+				metadataServiceUrlIPv6.Host = strings.Replace(metadataServiceUrlIPv6.Host, "iface", iface.Name, 1)
+				return nil
+			}
+		}
+	}
+	return fmt.Errorf("no active IPv6 network interface found")
+}
+
+// Fetches configuration from IPv4 and IPv6 metadata services
 func fetchConfigFromMetadataService(f *resource.Fetcher) ([]byte, error) {
-	res, err := f.FetchToBuffer(metadataServiceUrl, resource.FetchOptions{})
+	var response []byte
+	var err error
 
-	// the metadata server exists but doesn't contain any actual metadata,
-	// assume that there is no config specified
-	if err == resource.ErrNotFound {
```

Comment:
shouldn't we continue handling the not found error like it was done here?
I believe the same would need to happen for IPv6.